### PR TITLE
Adding additional tests for pbsfs on multi scheds

### PR DESCRIPTION
**Pull request by latha-subramanian**
#### Bug/feature Description
Additional automated tests for verifying pbsfs on multi scheds 

#### Affected Platform(s)
* Linux

#### Cause / Analysis / Design
Missing tests on some pbsfs's error conditions and "fairshare tree trim" on multi scheds

#### Solution Description
Added automated tests for pbsfs's error conditions and "fairshare tree trim" on multi scheds in pbs_multi_sched.py

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

[pbs_multi_sched_tests.log](https://github.com/PBSPro/pbspro/files/1764694/pbs_multi_sched_tests.log)

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_multi_sched.py [hunk 1]
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
...
+        try:
+            self.server.stop()
+            n = self.scheds['sc1'].query_fairshare()
+        except PbsFairshareError as e:
+            self.assertTrue(err_msg in e.msg)
+        self.server.start()
+
+    def test_pbfs_server_restart(self):
+        """
+        Test pbsfs -I <sched_name> when server is restarted
```
Comment (anamikau) [inline test/tests/functional/pbs_multi_sched.py]: update the description to "verify that server restart has no impact on fairshare usage"

File: test/tests/functional/pbs_multi_sched.py [hunk 2]
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
...
+        j1 = Job(TEST_USER, attrs={ATTR_queue: 'wq1'})
+        jid1 = self.server.submit(j1)
+        j2 = Job(TEST_USER1, attrs={ATTR_queue: 'wq1'})
+        jid2 = self.server.submit(j2)
+
+        t_start = int(time.time())
+        self.server.manager(MGR_CMD_SET, SCHED, {'scheduling': 'True'},
+                            id='sc1')
+        self.scheds[sched_name].log_match(
+            'Leaving Scheduling Cycle', starttime=t_start, max_attempts=10)
```
Comment (anamikau) [inline test/tests/functional/pbs_multi_sched.py]: remove max_attempts and use default (which is 30s)

File: test/tests/functional/pbs_multi_sched.py [hunk 3]
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
...
+        t_end = int(time.time())
+        job_list = self.scheds[sched_name].log_match(
+            'Considering job to run', starttime=t_start,
+            allmatch=True, max_attempts=10, endtime=t_end)
```
Comment (anamikau) [inline test/tests/functional/pbs_multi_sched.py]: again remove max_attempts and use default

File: test/tests/functional/pbs_multi_sched.py [hunk 4]
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
...
+        j1 = Job(TEST_USER, attrs={ATTR_queue: 'wq1'})
+        jid1 = self.server.submit(j1)
+        j2 = Job(TEST_USER1, attrs={ATTR_queue: 'wq1'})
+        jid2 = self.server.submit(j2)
+
+        t_start = int(time.time())
+        self.server.manager(MGR_CMD_SET, SCHED, {'scheduling': 'True'},
+                            id='sc1')
+        self.scheds[sched_name].log_match(
+            'Leaving Scheduling Cycle', starttime=t_start, max_attempts=10)
```
Comment (anamikau) [inline test/tests/functional/pbs_multi_sched.py]: remove max_attempts

File: test/tests/functional/pbs_multi_sched.py [hunk 5]
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
...
+        t_end = int(time.time())
+        job_list = self.scheds[sched_name].log_match(
+            'Considering job to run', starttime=t_start,
+            allmatch=True, max_attempts=10, endtime=t_end)
```
Comment (anamikau) [inline test/tests/functional/pbs_multi_sched.py]: remove max_attempts

**Comment by bhroam** [comment]
Codacy is complaining about a few unused variables.  Please remove those.

File: test/tests/functional/pbs_multi_sched.py [hunk 6]
```diff
@@ -699,3 +699,128 @@ def test_cmp_fairshare(self):
...
+        self.server.restart()
+        n = self.scheds['sc1'].query_fairshare().get_node(name=str(TEST_USER))
+        self.assertTrue(n.usage, 25)
+
+    def test_trim_fairshare(self):
+        """
+        Test pbsfs's trim fairshare tree functionality for multisched
+        """
+        self.common_setup()
+        sched_name = 'sc1'
```
Comment (visheshh) [inline test/tests/functional/pbs_multi_sched.py]: Does setup for test needs all scheduers ? else you can just setup_sc1 and queues and nodes as common_setup is heavy . 
Is there any intension of using sched_name variable than just using 'sc1' at all places ?

Comment (latha-subramanian) [reply test/tests/functional/pbs_multi_sched.py]: @visheshh . No. There wasn't any specific reason. I have reverted to 'sc1' as commonly done in other tests of the file.  Also, creating 'sc1' specific sched, queue and node. Got rid of common_setup(). Thanks!

**Comment by latha-subramanian** [comment]
@bhroam, @anamikau and @visheshh , I have addressed all of your review comments. Please have a look.  Thanks!

[pbs_multi_sched.log](https://github.com/PBSPro/pbspro/files/1768440/pbs_multi_sched.log)

File: test/tests/functional/pbs_multi_sched.py [hunk 7]
```diff
@@ -754,18 +742,30 @@ def test_pbfs_server_restart(self):
...
-        self.scheds[sched_name].set_sched_config({'fair_share': 'True'})
-        self.scheds[sched_name].set_fairshare_usage(TEST_USER, 100)
+        self.setup_sc1()
+        a = {'queue_type': 'execution',
+             'started': 'True',
+             'enabled': 'True'}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, id='wq1')
+
+        p1 = {'partition': 'P1'}
+        self.server.manager(MGR_CMD_SET, QUEUE, p1, id='wq1', expect=True)
```
Comment (bhroam) [inline test/tests/functional/pbs_multi_sched.py]: Combine the above two manager calls.  Put partition in the initial set of attributes.

File: test/tests/functional/pbs_multi_sched.py [hunk 8]
```diff
@@ -754,18 +742,30 @@ def test_pbfs_server_restart(self):
...
+        a = {'resources_available.ncpus': 2}
+        self.server.create_vnodes('vnode', a, 1, self.mom)
```
Comment (bhroam) [inline test/tests/functional/pbs_multi_sched.py]: Instead of creating one vnode, you could just use the mom directly.  Use a manager on self.mom.shortname

**Comment by latha-subramanian** [comment]
@bhroam , I have addressed your comments. Thanks!

[test_pbsfs_revert_to_defaults.log](https://github.com/PBSPro/pbspro/files/1776602/test_pbsfs_revert_to_defaults.log)

File: test/tests/functional/pbs_multi_sched.py [hunk 9]
```diff
@@ -751,14 +751,12 @@ def test_pbsfs_revert_to_defaults(self):
...
-        p1 = {'partition': 'P1'}
-        self.server.manager(MGR_CMD_SET, QUEUE, p1, id='wq1', expect=True)
+             'enabled': 'True',
+             'partition': 'P1'}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, id='wq1', expect=True)
         a = {'resources_available.ncpus': 2}
-        self.server.create_vnodes('vnode', a, 1, self.mom)
-        self.server.manager(MGR_CMD_SET, NODE, p1, id='vnode[0]', expect=True)
+        self.server.manager(MGR_CMD_SET, NODE, {'partition': 'P1'},
+                            id=self.mom.shortname, expect=True)
```
Comment (bhroam) [inline test/tests/functional/pbs_multi_sched.py]: Shouldn't this be partition and resources_available.ncpus?

Comment (latha-subramanian) [reply test/tests/functional/pbs_multi_sched.py]: Oops! I missed it.  Changed it now. Please review.
[test_pbsfs_revert_to_defaults.log](https://github.com/PBSPro/pbspro/files/1776656/test_pbsfs_revert_to_defaults.log)

**Comment by bhroam** [comment]
Please squash and rebase for the merge.

**Comment by latha-subramanian** [comment]
@bhroam,  I have squashed and rebased.  Please take it forward.

**Comment by bhroam** [comment]
I merged in the PR.

**Pull request merged**